### ISSUE TITLE
docker daemon hangs, does not respond to ps, kill  commands

### ISSUE THREAD

**Issue by r4um** [moby/moby#33093]
**Description**

Docker daemon hangs and does not respond to docker ps, kill commands

**Steps to reproduce the issue:**
1. Unable to reproduce

**Describe the results you received:**

docker ps, kill commands hang.

**Describe the results you expected:**

should respond with list of containers running / kill the containers.

**Additional information you deem important (e.g. issue happens only occasionally):**

Happens randomly, one thing observed in this case is, the pending signals in the
queue (`SigQ` in `/proc/PID/status`) are high ~7-31 across affected nodes.

Delivering a `SIGSTOP` and then immediately a `SIGCONT` recovers the docker daemon
process and it starts to respond to commands again.

go routines dump [docker.grd.txt](https://github.com/moby/moby/files/984733/docker.grd.txt)

**Output of `docker version`:**

```Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:47:50 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:47:50 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 10
 Running: 10
 Paused: 0
 Stopped: 0
Images: 42
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /data/aufs
 Backing Filesystem: xfs
 Dirs: 601
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 4.4.0-72-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 36
Total Memory: 58.97 GiB
Name: REDACTED
ID: 2GJ7:MN2A:JZJK:567M:NXF2:IKU2:77MK:RE2K:FXCR:G5T3:OD54:TSJR
Docker Root Dir: /data
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS EC2 c4.8xlarge node

**Comment by cpuguy83** [comment]
Thanks for reporting.

This is due to an issue with holding the container lock while waiting on exec stdio to complete, and additionally a sync issue with docker and containerd and never getting the signal that the stdio is completed.

Problem thread is here:
```
goroutine 190155 [semacquire, 5306 minutes]:
sync.runtime_Semacquire(0xc820ec228c)
	/usr/local/go/src/runtime/sema.go:43 +0x26
sync.(*WaitGroup).Wait(0xc820ec2280)
	/usr/local/go/src/sync/waitgroup.go:126 +0xb4
github.com/docker/docker/daemon.(*Daemon).StateChanged(0xc820464180, 0xc820358dc0, 0x40, 0x17657f0, 0xc, 0x100000000, 0xc820358e40, 0x40, 0x0, 0x0, ...)
	/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/monitor.go:62 +0xd8b
github.com/docker/docker/libcontainerd.(*container).handleEvent.func2()
	/usr/src/docker/.gopath/src/github.com/docker/docker/libcontainerd/container_linux.go:177 +0xa5
github.com/docker/docker/libcontainerd.(*queue).append.func1(0xc823ac7c01, 0xc8243b6b40, 0xc822c42820, 0xc82471efc0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/libcontainerd/queue_linux.go:26 +0x47
created by github.com/docker/docker/libcontainerd.(*queue).append
	/usr/src/docker/.gopath/src/github.com/docker/docker/libcontainerd/queue_linux.go:28 +0x1da
```

These are both fixed in later versions of Docker.
I recommend upgrading.

Closing since this is already resolved.
Thanks again!

**Comment by SaruKazen** [comment]
Can you please specify the version in which it is fixed, it seems to be occurring in the version 1.12 too

Output of docker version:
```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        
 OS/Arch:      linux/amd64
```

**Comment by cpuguy83** [comment]
@SaruKazen It was fixed since 1.13.0
